File: LEETCODE/ALL PROBLEMS/HARD/124. Binary Tree Maximum Path Sum.cpp
#include<bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int ans;
    int helper(TreeNode *root)
    {
        if (!root)
            return 0;
        int l = max(0, helper(root->left));
        int r = max(0, helper(root->right));
        ans = max(root->val + l + r, ans);
        return max(l, r) + root->val;
    }

    int maxPathSum(TreeNode *root)
    {
        ans = INT_MIN;
        helper(root);
        return ans;
    }
};

File: LEETCODE/HARD/124. Binary Tree Maximum Path  Sum.py
# Definition for a binary tree node.
import sys
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    ans = -sys.maxsize - 1
    def helper(self, root: TreeNode) -> int:
        if not root:
            return 0
        l = max(0, self.helper(root.left))
        r = max(0, self.helper(root.right))
        self.ans = max(self.ans, root.val + l + r)
        return max(l, r)+root.val

    def maxPathSum(self, root: TreeNode) -> int:
        self.helper(root)
        return self.ans

if __name__ == "__main__":
    vals = [1,2,3]
